Add `infrahubctl branch report` command

wvandeun opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Public API to fetch full diff trees with metadata
  * Branch report CLI command with detailed, table-based cleanup/status reports
  * Human-readable timestamp formatting and git-change detection in reports

* **Bug Fixes**
  * Always include properties and updated timestamps in node/relationship outputs
  * Expanded GraphQL variable handling to support nullable types and DateTime
  * Variable rendering now accepts date/time values

* **Tests**
  * Extended coverage for diff trees, branch reporting, and deep relationship processing

* **Documentation**
  * Added docs for the branch report command

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

wvandeun committed:
add optional graphql variable types and add datetime

ruff

wvandeun committed:
add get_diff_tree method to SDK client to get a diff object

wvandeun committed:
add updated_at meta data property for attributes and relationships

wvandeun committed:
add branch report command to infrahubctl

wvandeun committed:
fix

wvandeun committed:
add tests

wvandeun commented:
@coderabbitai review

wvandeun commented:
@coderabbitai review

wvandeun committed:
ruff

ogenstad commented on infrahub_sdk/graphql/constants.py:
```diff
@@ -1 +1,14 @@
-VARIABLE_TYPE_MAPPING = ((str, "String!"), (int, "Int!"), (float, "Float!"), (bool, "Boolean!"))
+from datetime import datetime
+
+VARIABLE_TYPE_MAPPING = (
+    (str, "String!"),
+    (str | None, "String"),
+    (int, "Int!"),
+    (int | None, "Int"),
+    (float, "Float!"),
+    (float | None, "Float"),
+    (bool, "Boolean!"),
+    (bool | None, "Boolean"),
+    (datetime, "DateTime!"),
+    (datetime | None, "DateTime"),
+)
```
If this PR targets `infrahub-develop` instead we've removed Python 3.9 support in that branch so it would be better to retarget this to avoid linting violations if we reintroduce `Union`.

wvandeun committed:
update documentation

wvandeun committed:
fix mypy

wvandeun committed:
fix: replace | union syntax with Union for Python 3.9 compatibility

Replace runtime | operator with Union from typing to support Python 3.9-3.13:
- graphql/constants.py: VARIABLE_TYPE_MAPPING tuples
- diff.py: GraphQL query variables dict

The | operator for types only works at runtime in Python 3.10+.

ogenstad commented on infrahub_sdk/graphql/constants.py:
```diff
@@ -1,14 +1,15 @@
+    (Union[bool, None], "Boolean"),
     (datetime, "DateTime!"),
-    (datetime | None, "DateTime"),
+    (Union[datetime, None], "DateTime"),
```
Hey Wim, I commented on the comment that CodeRabbit added. The better approach here would be to target the `infrahub-develop` branch where Python 3.9 is removed. So we don't want the old style `Union` here as it would cause linting issues in the other branch.

fatih-acar commented on .github/workflows/publish-pypi.yml:
```diff
@@ -41,39 +41,36 @@ jobs:
+      - name: Publish to PyPI
         if: ${{ inputs.publish }}
-        run: "poetry publish"
+        run: uv publish
```
Fixed in https://github.com/opsmill/infrahub-sdk-python/pull/639

wvandeun committed:
fix tests

wvandeun committed:
fail branch report command on main branch

wvandeun committed:
linting

wvandeun committed:
linting

wvandeun commented on infrahub_sdk/ctl/branch.py:
```diff
@@ -143,3 +185,115 @@ async def validate(branch_name: str, _: str = CONFIG_PARAM) -> None:
+            proposal_table.add_row("State", str(pc.state.value))  # type: ignore[union-attr]
+            proposal_table.add_row("Is draft", "Yes" if pc.is_draft.value else "No")  # type: ignore[union-attr]
+            proposal_table.add_row("Created by", pc.created_by.peer.name.value)  # type: ignore[union-attr]
+            proposal_table.add_row("Created at", format_timestamp(str(pc.created_by.updated_at)))  # type: ignore[union-attr]
```
This is a workaround that was implemented because we don't have easy access to the creation time of the proposed change in the SDK. This will change as soon as we implemented object level meta data in the 1.7 release of Infrahub.

ogenstad commented on infrahub_sdk/client.py:
```diff
@@ -1282,6 +1282,64 @@ async def get_diff_summary(
+        to_time: datetime | None = None,
+        timeout: int | None = None,
+        tracker: str | None = None,
+        raise_for_error: bool | None = None,
```
Will try to take a closer look tomorrow but `raise_for_error` is deprecated in the execute_graphql method. So we should not add that anywhere new.

wvandeun committed:
remove deprecated raise_for_error argument

wvandeun committed:
Merge remote-tracking branch 'origin/infrahub-develop' into wvd-20251114-infp388-branch-cleanup-mechanism

# Conflicts:
#	infrahub_sdk/node/constants.py
#	tests/unit/sdk/test_node.py

ogenstad commented on infrahub_sdk/client.py:
```diff
@@ -1282,6 +1282,62 @@ async def get_diff_summary(
 
         return node_diffs
 
+    async def get_diff_tree(
```
Do you see a value from an SDK point of view to have this as a method directly under the client? I don't have a strong objection to it but the client file is already massive as it is and at some point I'd like to break it apart into smaller components. Don't know if this command would remain in the same place at that point. Not really ready for that refactoring just now. But if we don't see a clear use case to have this within the SDK itself I'd suggest that we start out by just having it in the CTL.

wvandeun replied:
Yes I do, today there is no way to get this kind of diff for a single branch in the UI. The only way you do it today would be to craft a GraphQL query yourself. IE I do think this is a good intermediate to be able to expose this easier.

ogenstad commented on infrahub_sdk/client.py:
```diff
@@ -1282,6 +1282,62 @@ async def get_diff_summary(
+            num_added=diff_tree.get("num_added") or 0,
+            num_updated=diff_tree.get("num_updated") or 0,
+            num_removed=diff_tree.get("num_removed") or 0,
+            num_conflicts=diff_tree.get("num_conflicts") or 0,
```
Looking at this I think it comes from the fact that the GraphQL query itself has an incorrect return format. I.e. all of these should be required:

https://github.com/opsmill/infrahub/blob/infrahub-v1.5.3/backend/infrahub/graphql/queries/diff/tree.py#L62-L66

I.e. it should never return a `null` value like it looks like the query can do now. Based on the issue in Infrahub I think your approach would be the way to go. But we can create a followup task for it later to clean this up.

wvandeun replied:
Opened this task for this:
https://github.com/opsmill/infrahub/issues/7778

ogenstad commented on infrahub_sdk/ctl/branch.py:
```diff
@@ -18,6 +24,42 @@
+        dt = datetime.fromisoformat(timestamp.replace("Z", "+00:00"))
+        return dt.strftime("%Y-%m-%d %H:%M:%S")
+    except (ValueError, AttributeError):
+        return timestamp
```
This looks like a safeguard incase the response from Infrahub is invalid. With the docstring it's not clear that an incorrect value could be returned.

Is the issue here that we ignore typing somewhere and `timestamp` can be `None`?

ogenstad commented on infrahub_sdk/ctl/branch.py:
```diff
@@ -143,3 +185,115 @@ async def validate(branch_name: str, _: str = CONFIG_PARAM) -> None:
+            proposal_table.add_row("Created by", pc.created_by.peer.name.value)  # type: ignore[union-attr]
+            proposal_table.add_row("Created at", format_timestamp(str(pc.created_by.updated_at)))  # type: ignore[union-attr]
+            proposal_table.add_row("Approvals", str(len(pc.approved_by.peers)))  # type: ignore[union-attr]
+            proposal_table.add_row("Rejections", str(len(pc.rejected_by.peers)))  # type: ignore[union-attr]
```
We should avoid adding new ignore statements for typing. I'm guessing that these could be removed if we instead used the protocol definition and not the string `"CoreProposedChange"` when querying for these objects.

ogenstad commented on infrahub_sdk/ctl/branch.py:
```diff
@@ -143,3 +185,115 @@ async def validate(branch_name: str, _: str = CONFIG_PARAM) -> None:
+            proposal_table.add_row("Name", pc.name.value)  # type: ignore[union-attr]
+            proposal_table.add_row("State", str(pc.state.value))  # type: ignore[union-attr]
+            proposal_table.add_row("Is draft", "Yes" if pc.is_draft.value else "No")  # type: ignore[union-attr]
+            proposal_table.add_row("Created by", pc.created_by.peer.name.value)  # type: ignore[union-attr]
```
Based on the changes for immutable metadata in Infrahub 1.7, I'd consider waiting to include the creator of the PC.

ogenstad commented on infrahub_sdk/ctl/branch.py:
```diff
@@ -143,3 +185,115 @@ async def validate(branch_name: str, _: str = CONFIG_PARAM) -> None:
+
+    if branch_name == "main":
+        console.print("[red]Cannot create a report for the main branch!")
+        sys.exit(1)
```
While the main branch should be standard everywhere, the approach seems wrong here. I'd instead let the SDK query for the branch and then check the object on line 206 and if the returned object is the default branch raise an error there.

ogenstad commented on infrahub_sdk/ctl/branch.py:
```diff
@@ -143,3 +185,115 @@ async def validate(branch_name: str, _: str = CONFIG_PARAM) -> None:
+
+@app.command()
+@catch_exception(console=console)
+async def report(  # noqa: PLR0915
```
I'd prefer to now add ignore statements for complexity like this, we already have lots of them that we should clean up at some point. :( 

If we look at a snippet like this:
```python

    # Fetch diff tree (with metadata)
    diff_tree = await client.get_diff_tree(branch=branch_name)

    if diff_tree:
        branch_table.add_row("Diff last updated", format_timestamp(diff_tree["to_time"]))
        branch_table.add_row("Amount of additions", str(diff_tree["num_added"]))
        branch_table.add_row("Amount of deletions", str(diff_tree["num_removed"]))
        branch_table.add_row("Amount of updates", str(diff_tree["num_updated"]))
        branch_table.add_row("Amount of conflicts", str(diff_tree["num_conflicts"]))
    else:
        branch_table.add_row("Diff last updated", "No diff available")
        branch_table.add_row("Amount of additions", "-")
        branch_table.add_row("Amount of deletions", "-")
        branch_table.add_row("Amount of updates", "-")
        branch_table.add_row("Amount of conflicts", "-")
```

This looke like it could live in a function that would return a list of the rows that can them be added to the table.

ogenstad commented on infrahub_sdk/node/related_node.py:
```diff
@@ -64,7 +64,7 @@ def __init__(self, branch: str, schema: RelationshipSchemaAPI, data: Any | dict,
                 self._typename = node_data.get("__typename", None)
 
-            self.updated_at: str | None = data.get("updated_at", data.get("_relation__updated_at"))
+            self.updated_at: str | None = data.get("updated_at", properties_data.get("updated_at", None))
```
Was this intentional and a bug that needed to be fixed or do we have a response that looks like this from the backend?

wvandeun replied:
Yes, this is intentional. There was an issue where the `updated_at` metadata for an attribute and relationship was not properly populated.

wvandeun replied:
```python
import infrahub_sdk
print(infrahub_sdk.__version__)
1.15.1

from infrahub_sdk import InfrahubClientSync, Config
config = Config(api_token="06438eb2-8019-4776-878c-0941b1f1d1ec")
client = InfrahubClientSync(config=config,address="https://sandbox.infrahub.app")

device = client.get("InfraDevice", name__value="atl1-edge1", include=["interfaces"], prefetch_relationships=True, property=True)
print(device.name.updated_at) # None
print(device.interfaces.peers[0].updated_at) # None
```

ogenstad commented on infrahub_sdk/node/relationship.py:
```diff
@@ -87,9 +87,8 @@ def _generate_query_data(cls, peer_data: dict[str, Any] | None = None, property:
             for prop_name in PROPERTIES_OBJECT:
                 properties[prop_name] = {"id": None, "display_label": None, "__typename": None}
-
-        if properties:
```
Was this something that was required?

wvandeun replied:
Not necessary required. I just think this is cleaner. The previous approach was checking if we have value in the properties dict, whereas we are already checking if the properties should be queried using the property argument. This reads a bit cleaner, imo

ogenstad commented on tests/unit/ctl/test_branch_report.py:
```diff
@@ -0,0 +1,436 @@
+        method="GET",
+        url="http://mock/api/diff/files?branch=test-branch",
+        status_code=404,
+    )
```
I think this test is easier to read than the ones above, if we look at a fixture like `mock_git_files_empty_response` which is only used once and is so short I'd recommend having it inline within the test body as it's easier to see at a glance what's going on. If we know that we're going to use the fixture in multiple locations I think that's fine, but if they only are used once we also run the risk of leaving the fixture in place if the test is ever deleted.

ogenstad commented on tests/unit/ctl/test_branch_report.py:
```diff
@@ -0,0 +1,436 @@
+
+def test_branch_report_command_without_proposed_change(
+    mock_branch_report_command: HTTPXMock,
+    mock_schema_query_05,  # type: ignore[misc]
```
Looks like this is a `HTTPXMock` class to, so it should be fine to add that and remove the ignore statement?

ogenstad commented on tests/unit/sdk/test_diff_summary.py:
```diff
@@ -158,3 +161,239 @@ async def test_diffsummary(clients: BothClients, mock_diff_tree_query, client_ty
+
+    # Verify it's a Query object
+    assert hasattr(query, "render")
+    assert hasattr(query, "name")
```
Typing should catch this and if the name or render function doesn't exist the asserts just below would fail. So we can remove the hasattr checks.

ogenstad commented on infrahub_sdk/ctl/branch.py:
```diff
@@ -143,3 +185,115 @@ async def validate(branch_name: str, _: str = CONFIG_PARAM) -> None:
+    else:
+        branch_table.add_row("Git files changed", "N/A")
+
+    branch_table.add_row("Has schema changes", "Yes" if branch.has_schema_changes else "No")
```
I think this part could potentially be a bit misleading for stale branches. The only thing this flag does is to indicate if the schema differs compared to the default branch. This field is probably incorrectly named within Infrahub and should be called something like "schema in sync with default branch". The problem here would be if the default branch has moved on with other changes.

Could be that this naming would remain consistent with what we have within Infrahub, but I think the proper way to do this would be to do a diff of the schema as well. This might also be enough for an initial round.

wvandeun replied:
Ok, let's revisit this later.

wvandeun committed:
remove excpetion handling in format_timestamp

also add a better docstring

wvandeun committed:
remove typing ignore statements

wvandeun committed:
refactor creating branch report tables

wvandeun committed:
fix created_at metadata for proposed change in branch report

wvandeun committed:
update tests

wvandeun committed:
update tests

wvandeun committed:
update tests
